### PR TITLE
Initial website framework with accounts functionality

### PR DESCRIPTION
**Pull request by SirLegolot**
I just realized this is a huge chunk of code, which is my fault. The majority of the html files are almost identical (the navigation bar html code takes up the majority of the code for a file, the only things that are different is the content section of the file).

### CODE REVIEW

**Comment by SirLegolot** [comment]
@googlebot I fixed it.

**Comment by SirLegolot** [comment]
@googlebot I fixed it.

**Comment by SirLegolot** [comment]
@googlebot I fixed it.

File: project/src/main/webapp/businessAccount.html [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        </nav>
+      </div>
+      <main class="mdl-layout__content">
+        <div class="page-content" id="content">
+          <!-- Your content goes here -->
+          <h1 id="userGreeting">Welcome, business</h1>
+          <button id="logout" class="mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--accent">
+            Log out
+          </button>
+          <!-- TODO: Have a login and sign up servlet -->
```
Comment (OlagunjuPhillips) [inline project/src/main/webapp/businessAccount.html]: I think you can remove the TODO here

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  public CreateBusinessAccountServlet() {
+    super();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    userService = UserServiceFactory.getUserService();
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get required user information.
+    String userId = userService.getCurrentUser().getUserId();
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java:39]: Is there any chance of the current user not being created/initialized yet? Does something different need to happen in that case?

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    String userEmail = userService.getCurrentUser().getEmail();
+    String nickname = request.getParameter("nickname"); 
+    String street = request.getParameter("street");
+    String city = request.getParameter("city");
+    String state = request.getParameter("state");
+    String zipCode = request.getParameter("zipCode");
+    List<String> searchHistory = new ArrayList<>(); // no search history initally
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java]: Will you implement the search history for business accounts in addition to consumers? (It's fine either way, but I only remember seeing notes about implementing it for the consumer accounts).

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    // Set business information to defaults (as this is not a business).
+    boolean isUserBusinessOwner = false;
+    String businessId = null;
+    
+    // Create entity object that will be stored
+    Entity account = new Entity("Account", userId);
+    account.setProperty("userId", userId);
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java]: You could make this part more concise by replacing all the Strings you declared above with something like this:
`account.setProperty("userId", userService.getCurrentUser().getUserId());`

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    String userEmail = userService.getCurrentUser().getEmail();
+    String nickname = request.getParameter("nickname"); 
+    String street = request.getParameter("street");
+    String city = request.getParameter("city");
+    String state = request.getParameter("state");
+    String zipCode = request.getParameter("zipCode");
+    List<String> searchHistory = new ArrayList<>(); // no search history initally
+
+    // Set business information to defaults (as this is not a business).
+    boolean isUserBusinessOwner = false;
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java]: I think this should be set to true for this servlet (and update the comment and eventually set the businessId when you have it).

File: project/src/main/java/com/google/sps/servlets/CreateUserAccountServlet.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    String zipCode = request.getParameter("zipCode");
+    List<String> searchHistory = new ArrayList<>(); // no search history initally
+
+    // Set business information to defaults (as this is not a business).
+    boolean isUserBusinessOwner = false;
+    String businessId = null;
+    
+    // Create entity object that will be stored
+    Entity account = new Entity("Account", userId);
+    account.setProperty("userId", userId);
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/servlets/CreateUserAccountServlet.java]: Same comment as with the other servlet - you could replace all of the string declarations above by combining them with the calls to `account.setProperty()`

File: project/src/main/webapp/businessAccount.html [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                View my products
+              </button>
+              <button class="mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--colored">
+                Create a product set
+              </button>
+              <button class="mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--colored">
+                Add a product
+              </button>
+              <button class="mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect">
+                Delete Product/set
```
Comment (mariabajzek) [inline project/src/main/webapp/businessAccount.html]: Nit: you haven't capitalized "Product" on the other buttons. I think it's fine either way but should be made consistent.

File: project/src/main/webapp/businessAccount.html [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                <div class="mdl-textfield mdl-js-textfield">
+                  <input class="mdl-textfield__input" type="text" id="state" name="state">
+                  <label class="mdl-textfield__label" for="state">State</label>
+                </div>
+                <div class="mdl-textfield mdl-js-textfield">
+                  <input class="mdl-textfield__input" type="text" id="zipCode" name="zipCode">
+                  <label class="mdl-textfield__label" for="zipCode">Zip Code</label>
+                </div>
+                <br/><br/>
+                <button type="submit" class="mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--accent">
```
Comment (mariabajzek) [inline project/src/main/webapp/businessAccount.html:95]: Does something else need to be done to link this to the business account servlet? (I might have just missed it somewhere, or maybe you haven't gotten to it yet).

File: project/src/main/webapp/createUserAccount.html [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+          <!-- Your content goes here -->
+          <h1>Create a User Account</h1>
+          <!-- TODO: Have a login and sign up servlet -->
+          <form action="" method="POST" >
+            <p>Username*</p> 
+            <div class="mdl-textfield mdl-js-textfield">
+              <input class="mdl-textfield__input" type="text" id="nickname" name="nickname" required>
+              <label class="mdl-textfield__label" for="nickname">JohnDoe42</label>
+            </div><br/><br/>
+            <p>Address*</p>
```
Comment (mariabajzek) [inline project/src/main/webapp/createUserAccount.html:49]: You're collecting this so you can show the localized search results, right? It might be nice to at some point add a note about what it will be used for, so the user will know why you want to know.

**Review by mariabajzek — CHANGES_REQUESTED** [review]
This is a good start!

I think we can take a closer look at the HTML once more things are filled in and hooked up. I just did a high level overview of the HTML for now.

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java (hunk 2 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java:39]: Before I route to the createBusinessAccount page, I first direct the user to a login servlet that ensures that the user is logged in.

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java (hunk 3 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java]: I'm treating business accounts as normal users with extra permissions, so the person who owns the business account can browse the website normally as well.

File: project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java (hunk 5 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/servlets/CreateBusinessAccountServlet.java]: Oh you're right! I'll fix that

File: project/src/main/webapp/businessAccount.html (hunk 8 above)
Comment (SirLegolot) [reply project/src/main/webapp/businessAccount.html:95]: I currently do not have a business account servlet. I'm not sure if I need one. This page is loaded with information from the account servlet, which as of now is all I need.

File: project/pom.xml [hunk 10]
```diff
@@ -50,7 +50,7 @@
         <version>2.2.0</version>
         <configuration>
           <!-- TODO: set project ID. -->
-          <deploy.projectId>YOUR_PROJECT_ID_HERE</deploy.projectId>
+          <deploy.projectId>neelgandhi-step-2020</deploy.projectId>
```
Comment (mariabajzek) [inline project/pom.xml:53]: cloudberry-step-2020?

File: project/src/main/java/com/google/sps/data/Account.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+public class Account {
+	private String userId;
+	private String logoutUrl; 
+	private String nickname;
+	private String userEmail;
+	private boolean isUserBusinessOwner;
+	private List<String> searchHistory; 
+	private String street;
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/Account.java]: The indenting looks off here.

File: project/src/main/java/com/google/sps/data/Business.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
+package com.google.sps.data;
+import java.util.ArrayList;
+import java.util.List;
+
+public class Business {
+	private String businessId;
+  private String businessDisplayName;
+	private String street;
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/Business.java]: Indentation is off here.

File: project/src/main/java/com/google/sps/data/ServletLibrary.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  public static Account retrieveAccountInfo(DatastoreService datastore, UserService userService, String userId) {
+    Filter filter = new FilterPredicate("userId", FilterOperator.EQUAL, userId);
+    Query query = new Query("Account").setFilter(filter);
+    PreparedQuery pq = datastore.prepare(query);
+    Entity entity = pq.asSingleEntity();
+
+    if (entity == null) return null;
+
+    String logoutUrl = userService.createLogoutURL("/index.html");
+    String nickname = entity.getProperty("nickname").toString();
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/ServletLibrary.java]: I'm not familiar with the Entity type, but you might want to use `instanceof` to check these are Strings. It would be good to check the isUserBusinessOwner below as well.

File: project/src/main/java/com/google/sps/data/ServletLibrary.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    String userEmail = entity.getProperty("userEmail").toString();
+    boolean isUserBusinessOwner = (boolean) entity.getProperty("isUserBusinessOwner");
+
+    @SuppressWarnings("unchecked") // Documentation says to suppress warning this way
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/ServletLibrary.java:104]: Ah, is this to prevent it from complaining about the cast to ArrayList?

File: project/src/main/java/com/google/sps/data/VisionLibrary.java [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+      System.err.println("Error getting image labels: " + imageResponse.getError().getMessage());
+      return null;
+    }
+
+    return imageResponse;
+  }
+
+  public static String formatImageResponse(ImagesService imagesService, Gson gson, AnnotateImageResponse imageResponse, BlobKey blobKey) {
+    // These are labels for the overall image based on what the vision api thinks.
+    List<EntityAnnotation> labelAnnotations = imageResponse.getLabelAnnotationsList();
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/VisionLibrary.java:74]: You should check for null here, since it's possible imageResponse is null (see line 63 above).

File: project/src/main/java/com/google/sps/data/VisionLibrary.java [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+                                                 textInImage,
+                                                 cleanUpLabels(dominantColors),
+                                                 cleanUpLabels(objectsInImage), 
+                                                 cleanUpLabels(logosInImage)));
+  }
+
+  // This cleans up a imageLabel list by removing duplicates and blank labels.
+  // Also remove labels with a score of less than 40% (this can be adjusted later)
+  public static List<ImageLabel> cleanUpLabels(List<ImageLabel> labels) {
+    List<ImageLabel> cleanedList = new ArrayList<>();
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/VisionLibrary.java]: You could also consider using a set for this, and then it would do the deduplication for you.

File: project/src/main/java/com/google/sps/data/VisionLibrary.java [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+  
+  public static String getNearestColor(Color color) {
+    Color[] constantColors = new Color[] { Color.black, Color.blue, Color.cyan, Color.gray, Color.green, Color.magenta, Color.orange, Color.pink, Color.red, Color.white, Color.yellow };
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/VisionLibrary.java]: It would be good to break this up (maybe in the middle) so it fits within 80 characters.

File: project/src/main/java/com/google/sps/data/VisionLibrary.java [hunk 18]
```diff
@@ -0,0 +1,330 @@
...
+  // the main idea is that the closest color visually is not necessarily the closest
+  // color in terms of a strict euclidean distance.
+  public static int colorDistance(Color c1, Color c2) {
+    // int red1 = c1.getRed();
+    // int red2 = c2.getRed();
+    // int rmean = (red1 + red2) >> 1;
+    // int r = red1 - red2;
+    // int g = c1.getGreen() - c2.getGreen();
+    // int b = c1.getBlue() - c2.getBlue();
+    // return (((512+rmean)*r*r)>>8) + 4*g*g + (((767-rmean)*b*b)>>8);
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/VisionLibrary.java:315]: If this was here for a test that you don't need anymore, it would be good to remove it. You can leave it for now if you're going to use it for more testing though.

File: project/src/main/java/com/google/sps/data/VisionLibrary.java [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+    int b = c1.getBlue() - c2.getBlue();
+    return r*r + g*g + b*b;
+  }  
+
+  public static Color convertToJavaColor(com.google.type.Color color) {
+    return new Color(Math.round(color.getRed()), Math.round(color.getGreen()), Math.round(color.getBlue()));
+  }
+
+  public static String getNearestColorString(Color color) {
+    if (color.equals(Color.black)) return "black";
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/VisionLibrary.java:329]: You can replace this with a switch statement to make it a bit cleaner.

https://www.w3schools.com/java/java_switch.asp

File: project/src/main/webapp/createProductSet.html [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+          <a class="mdl-navigation__link" href="">About</a>
+          <a class="mdl-navigation__link" href="/login">Account</a>
+        </nav>
+      </div>
+      <main class="mdl-layout__content">
+        <div class="page-content" id="content">
+          <!-- Your content goes here -->
+          <h1>Create a Product Set</h1>
+          <form action="/createProductSet" method="POST">
+            <p>Select from the drop down below or typing a new set.</p>
```
Comment (mariabajzek) [inline project/src/main/webapp/createProductSet.html]: Maybe rephrase to "...or type a new set name."?

**Review by mariabajzek — CHANGES_REQUESTED** [review]

File: project/pom.xml (hunk 10 above)
Comment (SirLegolot) [reply project/pom.xml:53]: I put my project there for now as having multiple people deploying to cloudberry when testing at the same time will cause problems

File: project/src/main/java/com/google/sps/data/Account.java (hunk 11 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/data/Account.java]: How exactly is the indenting off? Everything is indented 2 spaces

File: project/src/main/java/com/google/sps/data/ServletLibrary.java (hunk 14 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/data/ServletLibrary.java:104]: yup

File: project/src/main/java/com/google/sps/data/VisionLibrary.java (hunk 16 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/data/VisionLibrary.java]: I would need to defined my own hashCode() and equals() methods for the imageLabel object before I can use hashsets made of imageLabel objects. That way, it would properly recognize two objects as "equal" (having the same description). I guess it would be cleaner to do it that way, but I had chosen this method as it is slightly simpler. Is there a good way to write a hash function for my ImageLabel class?

File: project/src/main/java/com/google/sps/data/VisionLibrary.java (hunk 19 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/data/VisionLibrary.java:329]: From my brief search online, it seems that case statements don't work except for primitive types such as strings? So I can't do something like this because it doesn't know equality:
switch (color) {
    case Color.black {return "black"; break;}
    ...
}
Not exactly sure, but what would be a good way to approach this?

File: project/pom.xml (hunk 10 above)
Comment (mariabajzek) [reply project/pom.xml:53]: Ok, makes sense.

File: project/src/main/java/com/google/sps/data/Account.java (hunk 11 above)
Comment (mariabajzek) [reply project/src/main/java/com/google/sps/data/Account.java]: Mine is showing lines 6-12 being indented more than lines 13-15. It's possible it's just a weird display issue though. I'm not sure how to tell.

Comment (SirLegolot) [reply project/src/main/java/com/google/sps/data/Account.java]: Oh that's interesting haha. It shows incorrect indenting on github for me as well, but on my code in the editor it looks fine. I might have changed it after I pushed this code.

File: project/src/main/java/com/google/sps/data/VisionLibrary.java (hunk 19 above)
Comment (mariabajzek) [reply project/src/main/java/com/google/sps/data/VisionLibrary.java:329]: Ah, you're right, sorry about that. I hadn't noticed the type of `color`.

In that case, I think what you have is already good.

File: project/src/main/java/com/google/sps/data/VisionLibrary.java (hunk 16 above)
Comment (mariabajzek) [reply project/src/main/java/com/google/sps/data/VisionLibrary.java]: It's not uncommon to define your own equals() and hashCode() type things (not only in Java, you'll probably come across it in C++ and other languages as well).

If you only care about the description and not the score, your hashcode can just be the hash of the description. That way you might get some undefined behavior if you have to compare two ImageLabels with the same description and different scores, although I think that wouldn't happen in this case?

File: project/src/main/java/com/google/sps/data/ServletLibrary.java [hunk 21]
```diff
@@ -42,18 +65,45 @@ public static Account retrieveAccountInfo(DatastoreService datastore, UserServic
...
+        (_zipCode instanceof String)) {
+      nickname = _nickname.toString();
+      userEmail = _userEmail.toString();
+      isUserBusinessOwner = (boolean) _isUserBusinessOwner;
+      street = _street.toString();
+      city = _city.toString();
+      state = _state.toString();
+      zipCode = _zipCode.toString();
+    } else {
+      System.err.println("Entity properties are of an incorrect type.");
```
Comment (mariabajzek) [inline project/src/main/java/com/google/sps/data/ServletLibrary.java:99]: This would be easier to debug in the future if you broke it down so you could log the specific entity that was the wrong type. I'm fine with it this way though, just wanted to point it out for a possible improvement.

**Review by mariabajzek — APPROVED** [review]
Looks good, just one optional suggestion.

File: project/src/main/java/com/google/sps/data/ServletLibrary.java (hunk 21 above)
Comment (SirLegolot) [reply project/src/main/java/com/google/sps/data/ServletLibrary.java:99]: True I could do that! I chose not to do that for now as it would require having like 30 different if statements, but it is definitely good to have it that way for debugging quicker.
